File: bots/intgrah/v56_py/util/debug.py
"""
Bot-side debug helpers — structured per-turn JSON tree + cambc indicator
overlays.

Per-turn flow:
- The first `Scope` of the turn starts a fresh root node and stack.
- Nested `Scope`s become child nodes of the current top-of-stack scope.
- Bodies append `msg` and `vis` nodes via `debug()` and `vis()`.
- At end of turn, the bot calls `flush()` which prints the root tree as
  one line of JSON to stdout.

State lives in a `DebugCtx` struct with one process-global instance. Each
cdylib has its own copy of the static, and the engine serialises bot
calls per turn, so no locking is needed in practice — `unsafe` reads from
a `static mut` are fine on this single-threaded path.
"""
from __future__ import annotations

from typing import Final
import time
import json

def _pyrust_install_aliases() -> None:
    try:
        import cambc as _c
    except ImportError:
        return
    for _name in ("Team", "ResourceType", "EntityType", "Environment", "Direction"):
        _cls = getattr(_c, _name, None)
        if _cls is None:
            continue
        for _m in list(_cls):
            _p = "".join(_s.capitalize() for _s in _m.name.split("_"))
            if not hasattr(_cls, _p):
                setattr(_cls, _p, _m)
    # Hot-path: cache (dx, dy) per Direction so `pos.add(d)` skips the
    # 9-entry dict construction in `Direction.delta`. ~3.3x speedup.
    _Pos = getattr(_c, "Position", None)
    _Dir = getattr(_c, "Direction", None)
    if _Pos is not None and _Dir is not None:
        _DELTA = {_d: _d.delta() for _d in _Dir}
        def _fast_add(self, d, _DELTA=_DELTA, _Pos=_Pos):
            _dx, _dy = _DELTA[d]
            return _Pos(self.x + _dx, self.y + _dy)
        _Pos.add = _fast_add
        def _fast_delta(self, _DELTA=_DELTA):
            return _DELTA[self]
        _Dir.delta = _fast_delta
    # Hot-path: pre-build all 5000 Position objects so `pos_of(p)`
    # is a tuple __getitem__ instead of a fresh `Position(x, y)` call.
    if _Pos is not None:
        _POS_TABLE = tuple(_Pos(x=_i % 100, y=_i // 100) for _i in range(5000))
        _Pos.lookup = _POS_TABLE.__getitem__
    # Hot-path: precomputed PosInt distance tables. Indexed by
    # `p - q + DIST_OFFSET` where DIST_OFFSET=4949, length 9899.
    if _Pos is not None:
        _STRIDE = 100
        _DIST_OFFSET = 4949
        _DIST_LEN = 9899
        _MAX_D = 49
        _ds = [0] * _DIST_LEN
        _mh = [0] * _DIST_LEN
        _ch = [0] * _DIST_LEN
        for _dy in range(-_MAX_D, _MAX_D + 1):
            for _dx in range(-_MAX_D, _MAX_D + 1):
                _i = _dy * _STRIDE + _dx + _DIST_OFFSET
                _adx = _dx if _dx >= 0 else -_dx
                _ady = _dy if _dy >= 0 else -_dy
                _ds[_i] = _dx * _dx + _dy * _dy
                _mh[_i] = _adx + _ady
                _ch[_i] = _adx if _adx > _ady else _ady
        _Pos.dist_sq_table = tuple(_ds).__getitem__
        _Pos.manhat_table = tuple(_mh).__getitem__
        _Pos.chebyshev_table = tuple(_ch).__getitem__
_pyrust_install_aliases()
del _pyrust_install_aliases

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from cambc import Controller, ControllerApi, Position
from config import DEBUG_LOG
from util.visualiser import Dumper
if TYPE_CHECKING:
    from util.visualiser import Dump, DumpBoolGrid, DumpU8Grid, DumpI16Grid, DumpU16Grid, DumpF32Grid, DumpTiles, DumpTile, DumpDot, DumpPath, DumpVectorField, DumpScalar
TYPE_KEY: Final[str] = "$type"
"""Discriminator key for typed JSON nodes (matches Python `_TYPE = "$type"`)."""

class Frame:
    """
    Per-frame entry: the index of this scope's child slot inside its parent's
    `children` array, plus (for timed scopes) the start nanosecond timestamp.
    The frame at index 0 has `parent_child_idx = None` because it is the root.
    """
    parent_child_idx: int | None
    t0_ns: int | None

    def __init__(self, parent_child_idx: int | None, t0_ns: int | None):
        self.parent_child_idx = parent_child_idx
        self.t0_ns = t0_ns

class DebugCtx:
    """
    Process-global debug state. The `Dumper`'s same_cache is keyed by
    `(current_bot_id, name)` so multiple builders running through the
    same static still get per-bot same-elision.
    """
    root: Value | None
    frames: list[Frame]
    dumper: Dumper
    last_flush_us: int
    current_bot_id: int

    def __init__(self):
        self.root = None
        self.frames = []
        self.dumper = Dumper()
        self.last_flush_us = 0
        self.current_bot_id = -1

    def current_scope_mut(self):
        """
        Walk into the current scope node (the deepest open scope), using
        the parent-child indices recorded in `frames`.
        """
        root = self.root
        node: Value = root
        for f in self.frames[1:]:
            idx = f.parent_child_idx
            node = node["children"][idx]
        return node

    def push_scope(self, label, timed):
        node = {"$type": "scope", "name": str(label), "children": []}
        t0_ns = time.perf_counter_ns() if timed else None
        if (self.root is None):
            self.root = node
            self.frames.append(Frame(parent_child_idx=None, t0_ns=t0_ns))
            return
        parent = self.current_scope_mut()
        children = parent["children"]
        idx = len(children)
        children.append(node)
        self.frames.append(Frame(parent_child_idx=idx, t0_ns=t0_ns))

    def pop_scope(self):
        frame = (self.frames.pop() if self.frames else None)
        t0_ns = frame.t0_ns
        if t0_ns is not None:
            us = (time.perf_counter_ns() - t0_ns) // 1000
            if (not self.frames):
                root = self.root
                root["us"] = us
            else:
                idx = frame.parent_child_idx
                parent = self.current_scope_mut()
                parent["children"][idx]["us"] = us
        if (not self.frames):
            self.root = None

    def emit_child(self, node):
        if (not self.frames):
            return
        parent = self.current_scope_mut()
        parent["children"].append(node)

    def debug(self, tmpl, args):
        node = {"$type": "msg", "tmpl": str(tmpl), "args": args}
        self.emit_child(node)

    def vis(self, name, value):
        if (not self.frames):
            return
        root = self.root
        node: Value = root
        for f in self.frames[1:]:
            idx = f.parent_child_idx
            node = node["children"][idx]
        children = node["children"]
        self.dumper.dump(children, self.current_bot_id, name, value)

    def flush(self):
        prev_us = self.last_flush_us
        root = self.root
        root["prev_flush_us"] = prev_us
        t0_ns = time.perf_counter_ns()
        payload = json.dumps(root)
        print(f"{payload}")
        self.last_flush_us = (time.perf_counter_ns() - t0_ns) // 1000

    @staticmethod
    def default():
        return DebugCtx()
CTX: DebugCtx | None = None
"""
Process-global debug context. The `Dumper` inside same-elides per
`(bot_id, name)`, so multiple builders sharing the static still get
per-builder same_cache isolation. `current_bot_id` is set at the top
of `Player::run` via `set_current_bot` and used by `Dumper.dump`.
"""

def ctx():
    global CTX
    if (CTX is None):
        CTX = DebugCtx()
    return CTX

def set_current_bot(id):
    """
    Set the bot id used as the same-elision cache key for subsequent
    `vis()` calls. Call at the top of each `run()` before any scope opens.
    """
    if not DEBUG_LOG:
        return
    ctx().current_bot_id = id

class Scope:

    def __init__(self, label=None):
        self.label = str(label) if label is not None else ""
        if DEBUG_LOG:
            ctx().push_scope(self.label, False)

    def __enter__(self):
        return self

    def __exit__(self, *_a):
        if DEBUG_LOG:
            ctx().pop_scope()

    @staticmethod
    def new(label):
        """Push an untimed scope onto the stack. The returned guard pops on drop."""
        ctx().push_scope(label, False)
        __self = Scope.__new__(Scope)
        __self.label = str(label)
        return __self
    @staticmethod
    def new_timed(label):
        """
        Push a timed scope; on drop, records `us` (microseconds elapsed).
        Mirrors Python `Scope(label, time=True)`.
        """
        ctx().push_scope(label, True)
        __self = Scope.__new__(Scope)
        __self.label = str(label)
        return __self
    def drop(self):
        ctx().pop_scope()

def debug(tmpl, args):
    """
    Append a `msg` node under the current scope. `tmpl` is a Python-style
    format-string fragment using `{name}` slots; `args` provide the values
    referenced by those slots.
    """
    if not DEBUG_LOG:
        return
    ctx().debug(tmpl, args)

def vis(name, value):
    """
    Append a vis node under the current scope, routed through the per-unit
    `Dumper` for same-elision.
    """
    if not DEBUG_LOG:
        return
    ctx().vis(name, value)

def flush():
    """
    Print the root scope as one JSON line to stdout. MUST be called from
    inside the top-level `Scope::new("turn")` block, before that block's
    guard drops.
    """
    if not DEBUG_LOG:
        return
    ctx().flush()

def dot(ct, pos, r, g, b):
    """
    Wrapper over `Controller::draw_indicator_dot`. Engine-side overlay,
    visible to all spectators, on/off globally per replay.
    """
    if not DEBUG_LOG:
        return
    ct.draw_indicator_dot(pos, r, g, b)

def line(ct, pos_a, pos_b, r, g, b):
    """Wrapper over `Controller::draw_indicator_line`."""
    if not DEBUG_LOG:
        return
    ct.draw_indicator_line(pos_a, pos_b, r, g, b)

File: bots/rs/v56/src/util/debug.rs
//! Bot-side debug helpers — structured per-turn JSON tree + cambc indicator
//! overlays.
//!
//! Per-turn flow:
//! - The first `Scope` of the turn starts a fresh root node and stack.
//! - Nested `Scope`s become child nodes of the current top-of-stack scope.
//! - Bodies append `msg` and `vis` nodes via `debug()` and `vis()`.
//! - At end of turn, the bot calls `flush()` which prints the root tree as
//!   one line of JSON to stdout.
//!
//! State lives in a `DebugCtx` struct with one process-global instance. Each
//! cdylib has its own copy of the static, and the engine serialises bot
//! calls per turn, so no locking is needed in practice — `unsafe` reads from
//! a `static mut` are fine on this single-threaded path.

use cambc::{Controller, ControllerApi, Position};
use serde_json::{Map, Value};

use crate::config::DEBUG_LOG;
use crate::util::visualiser::{Dump, Dumper};

#[pyrust::inline]
/// Discriminator key for typed JSON nodes (matches Python `_TYPE = "$type"`).
const TYPE_KEY: &str = "$type";

/// Per-frame entry: the index of this scope's child slot inside its parent's
/// `children` array, plus (for timed scopes) the start nanosecond timestamp.
/// The frame at index 0 has `parent_child_idx = None` because it is the root.
struct Frame {
    parent_child_idx: Option<usize>,
    t0_ns: Option<u64>,
}

/// Process-global debug state. The `Dumper`'s same_cache is keyed by
/// `(current_bot_id, name)` so multiple builders running through the
/// same static still get per-bot same-elision.
pub struct DebugCtx {
    /// The root node of the current turn's tree. `None` when no scope is
    /// active. Owns all nested children.
    root: Option<Value>,
    /// Frame stack. Empty when no scope is active. The top frame is always
    /// the current scope; new children are appended to it.
    frames: Vec<Frame>,
    /// Per-bot same-elision dumper for `vis()`.
    dumper: Dumper,
    /// Microseconds spent inside the previous `flush()` call. Recorded into
    /// the next root as `prev_flush_us` so the visualiser can show I/O cost.
    last_flush_us: u64,
    /// The id of the bot currently running. Set by `set_current_bot` at
    /// the top of each `Player::run`. Used by `Dumper` as part of the
    /// same_cache key.
    pub current_bot_id: i32,
}

impl DebugCtx {
    #[must_use]
    pub fn new() -> Self {
        Self {
            root: None,
            frames: pyrust::vec::new!(),
            dumper: Dumper::new(),
            last_flush_us: 0,
            current_bot_id: -1,
        }
    }

    /// Walk into the current scope node (the deepest open scope), using
    /// the parent-child indices recorded in `frames`.
    fn current_scope_mut(&mut self) -> &mut Value {
        let root = pyrust::expect!(pyrust::as_mut!(self.root), "scope active but root is None");
        let mut node: &mut Value = root;
        // Skip frame 0 (the root frame). Each subsequent frame says which
        // child slot of the previous scope it occupies.
        for f in &self.frames[1..] {
            let idx = pyrust::expect!(f.parent_child_idx, "non-root frame must have idx");
            node = &mut node["children"][idx];
        }
        node
    }

    pub fn push_scope(&mut self, label: &str, timed: bool) {
        let node = serde_json::json!({
            TYPE_KEY: "scope",
            "name": pyrust::to_string!(label),
            "children": [],
        });
        let t0_ns = if timed {
            Some(pyrust::time::now_ns!())
        } else {
            None
        };
        if pyrust::is_none!(self.root) {
            // First scope of the turn: this becomes the root.
            self.root = Some(node);
            pyrust::vec::push!(
                self.frames,
                Frame {
                    parent_child_idx: None,
                    t0_ns,
                }
            );
            return;
        }
        let parent = self.current_scope_mut();
        let children = pyrust::serde::array_mut!(parent["children"]);
        let idx = pyrust::len!(children);
        pyrust::vec::push!(children, node);
        pyrust::vec::push!(
            self.frames,
            Frame {
                parent_child_idx: Some(idx),
                t0_ns,
            }
        );
    }

    pub fn pop_scope(&mut self) {
        let frame = pyrust::expect!(
            pyrust::vec::pop!(self.frames),
            "Scope::drop with empty frame stack"
        );
        if let Some(t0_ns) = frame.t0_ns {
            let us = (pyrust::time::now_ns!() - t0_ns) / 1000;
            if pyrust::vec::is_empty!(self.frames) {
                // The frame we just popped was the root.
                let root = pyrust::expect!(pyrust::as_mut!(self.root), "ROOT must be Some");
                root["us"] = serde_json::Value::Number(pyrust::into!(us));
            } else {
                let idx = pyrust::expect!(frame.parent_child_idx, "non-root has idx");
                let parent = self.current_scope_mut();
                parent["children"][idx]["us"] = serde_json::Value::Number(pyrust::into!(us));
            }
        }
        // Once the outer scope drops, clear ROOT so the next turn's first
        // `Scope::new` rebuilds a fresh tree. Mirrors Python's `_stack`
        // emptying on outer-scope exit.
        if pyrust::vec::is_empty!(self.frames) {
            self.root = None;
        }
    }

    fn emit_child(&mut self, node: Value) {
        if pyrust::vec::is_empty!(self.frames) {
            return;
        }
        let parent = self.current_scope_mut();
        pyrust::vec::push!(pyrust::serde::array_mut!(parent["children"]), node);
    }

    pub fn debug(&mut self, tmpl: &str, args: Map<String, Value>) {
        let node = serde_json::json!({
            TYPE_KEY: "msg",
            "tmpl": pyrust::to_string!(tmpl),
            "args": Value::Object(args),
        });
        self.emit_child(node);
    }

    pub fn vis(&mut self, name: &str, value: &Dump) {
        if pyrust::vec::is_empty!(self.frames) {
            return;
        }
        // Split-borrow: walk root → child slot via raw indexing so the
        // borrow into the children Vec is disjoint from `self.dumper`.
        let root = pyrust::expect!(pyrust::as_mut!(self.root), "scope active but root is None");
        let mut node: &mut Value = root;
        for f in &self.frames[1..] {
            let idx = pyrust::expect!(f.parent_child_idx, "non-root frame must have idx");
            node = &mut node["children"][idx];
        }
        let children = pyrust::serde::array_mut!(node["children"]);
        self.dumper.dump(children, self.current_bot_id, name, value);
    }

    pub fn flush(&mut self) {
        let prev_us = self.last_flush_us;
        let root = pyrust::expect!(
            pyrust::as_mut!(self.root),
            "flush() called outside any Scope"
        );
        root["prev_flush_us"] = serde_json::Value::Number(pyrust::into!(prev_us));
        let t0_ns = pyrust::time::now_ns!();
        let payload = pyrust::expect!(serde_json::to_string(root), "root scope must serialise");
        println!("{payload}");
        self.last_flush_us = (pyrust::time::now_ns!() - t0_ns) / 1000;
    }
}

impl Default for DebugCtx {
    fn default() -> Self {
        Self::new()
    }
}

/// Process-global debug context. The `Dumper` inside same-elides per
/// `(bot_id, name)`, so multiple builders sharing the static still get
/// per-builder same_cache isolation. `current_bot_id` is set at the top
/// of `Player::run` via `set_current_bot` and used by `Dumper.dump`.
static mut CTX: Option<DebugCtx> = None;

#[allow(static_mut_refs)]
fn ctx() -> &'static mut DebugCtx {
    if unsafe { pyrust::is_none!(CTX) } {
        unsafe { CTX = Some(DebugCtx::new()) };
    }
    unsafe { pyrust::unwrap!(pyrust::as_mut!(CTX)) }
}

/// Set the bot id used as the same-elision cache key for subsequent
/// `vis()` calls. Call at the top of each `run()` before any scope opens.
pub fn set_current_bot(id: i32) {
    if !DEBUG_LOG {
        return;
    }
    ctx().current_bot_id = id;
}

/// Tree-internal scope guard. The full implementation only exists when
/// `DEBUG_LOG=1` was set at build time (build.rs emits `--cfg debug_log`).
/// When the cfg is off, `Scope` is a ZST whose constructors and Drop are
/// no-ops, so `let _g = Scope::new(...)` compiles to nothing in release.
///
/// pyrust always sees the `#[pyrust::context_manager]` annotation on this
/// struct (it ignores `#[cfg]`), so Python emission still gets `with`-blocks
/// regardless of the native build's cfg state.
#[pyrust::context_manager]
#[cfg(debug_log)]
pub struct Scope {
    pub label: String,
}

#[cfg(debug_log)]
impl Scope {
    /// Push an untimed scope onto the stack. The returned guard pops on drop.
    #[must_use]
    pub fn new(label: &str) -> Self {
        ctx().push_scope(label, false);
        Self {
            label: pyrust::to_string!(label),
        }
    }

    /// Push a timed scope; on drop, records `us` (microseconds elapsed).
    /// Mirrors Python `Scope(label, time=True)`.
    #[must_use]
    pub fn new_timed(label: &str) -> Self {
        ctx().push_scope(label, true);
        Self {
            label: pyrust::to_string!(label),
        }
    }
}

#[cfg(debug_log)]
impl Drop for Scope {
    fn drop(&mut self) {
        ctx().pop_scope();
    }
}

#[pyrust::context_manager]
#[cfg(not(debug_log))]
pub struct Scope;

#[cfg(not(debug_log))]
impl Scope {
    #[inline(always)]
    #[must_use]
    pub const fn new(_label: &str) -> Self {
        Self
    }

    #[inline(always)]
    #[must_use]
    pub const fn new_timed(_label: &str) -> Self {
        Self
    }
}

/// Append a `msg` node under the current scope. `tmpl` is a Python-style
/// format-string fragment using `{name}` slots; `args` provide the values
/// referenced by those slots.
pub fn debug(tmpl: &str, args: Map<String, Value>) {
    if !DEBUG_LOG {
        return;
    }
    ctx().debug(tmpl, args);
}

/// Append a vis node under the current scope, routed through the per-unit
/// `Dumper` for same-elision.
pub fn vis(name: &str, value: &Dump) {
    if !DEBUG_LOG {
        return;
    }
    ctx().vis(name, value);
}

/// Print the root scope as one JSON line to stdout. MUST be called from
/// inside the top-level `Scope::new("turn")` block, before that block's
/// guard drops.
pub fn flush() {
    if !DEBUG_LOG {
        return;
    }
    ctx().flush();
}

/// Wrapper over `Controller::draw_indicator_dot`. Engine-side overlay,
/// visible to all spectators, on/off globally per replay.
pub fn dot(ct: &mut Controller<'_>, pos: Position, r: i32, g: i32, b: i32) {
    if !DEBUG_LOG {
        return;
    }
    pyrust::unwrap!(ct.draw_indicator_dot(pos, r, g, b));
}

/// Wrapper over `Controller::draw_indicator_line`.
pub fn line(ct: &mut Controller<'_>, pos_a: Position, pos_b: Position, r: i32, g: i32, b: i32) {
    if !DEBUG_LOG {
        return;
    }
    pyrust::unwrap!(ct.draw_indicator_line(pos_a, pos_b, r, g, b));
}
